### PR TITLE
[VL] Velox splitter supports dictionary encoding

### PR DESCRIPTION
**Pull request by zuochunwei**
## What changes were proposed in this pull request?

You can review and approve this PR.

but please don't merge this PR before @jinchengchenghh enable the Velox splitter PR.

I will merge it myself later.

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp/velox/shuffle/VeloxSplitter.cc [hunk 1]
```diff
@@ -365,115 +366,34 @@ arrow::Status VeloxSplitter::SplitFixedWidthValueBuffer(const velox::RowVector&
   for (auto col = 0; col < fixed_width_column_count_; ++col) {
     auto col_idx = simple_column_indices_[col];
     auto column = rv.childAt(col_idx);
-    assert(column->isFlatEncoding());
 
-    const uint8_t* src_addr = (const uint8_t*)column->valuesAsVoid();
+    velox::DecodedVector dv(*column);
```
Comment (jinchengchenghh) [inline cpp/velox/shuffle/VeloxSplitter.cc:370]: It is suggested to save `velox::DecodedVector dv`, and then invoke dv.decode(*column) for each column

**Comment by jinchengchenghh** [comment]
Other is OK, `valueAt` may need optimize.

File: cpp/velox/shuffle/VeloxSplitter.cc (hunk 1 above)
Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxSplitter.cc]: I used another way

**Comment by zuochunwei** [comment]
> Other is OK, `valueAt` may need optimize.

yes, `valueAt()` is a virtual function call

File: cpp/velox/shuffle/VeloxSplitter.cc (hunk 1 above)
Comment (jinchengchenghh) [reply cpp/velox/shuffle/VeloxSplitter.cc]: In this way, we will initialize class `velox::DecodedVector` with each Batch, but for velox purpose, We should create one `velox::DecodedVector`  for all Batches, https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/ApproxDistinctAggregate.cpp#L377

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxSplitter.cc]: OK

File: cpp/velox/shuffle/VeloxSplitter.h [hunk 2]
```diff
@@ -385,6 +390,9 @@ class VeloxSplitter {
 
   // shared by all partition writers
   std::shared_ptr<arrow::ipc::IpcPayload> schema_payload_;
+
+  facebook::velox::DecodedVector decoded_vector_;
```
Comment (jinchengchenghh) [inline cpp/velox/shuffle/VeloxSplitter.h:394]: Each column should keep one `DecodedVector` as here do https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/ApproxDistinctAggregate.cpp#L377

**Comment by jinchengchenghh** [comment]
Can you add a unit test for one splitter?

**Comment by zuochunwei** [comment]
> Can you add a unit test for one splitter?

OK

File: cpp/velox/shuffle/VeloxSplitter.h (hunk 2 above)
Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxSplitter.h:394]: the DecodedVector is used to access  the wrapped Vector's data for the purpose of performance, and the `decode()` will call `reset()`, it's just a shell and `decode()` will reuse it.

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxSplitter.h:394]: you can check its help document [DecodeVector](https://facebookincubator.github.io/velox/develop/dictionary-encoding.html#decodedvector)

Comment (jinchengchenghh) [reply cpp/velox/shuffle/VeloxSplitter.h:394]: https://facebookincubator.github.io/velox/develop/aggregate-functions.html

**Comment by jinchengchenghh** [comment]
Wait for https://github.com/oap-project/gluten/pull/1077

**Comment by zuochunwei** [comment]
> In velox document https://facebookincubator.github.io/velox/develop/aggregate-functions.html, and velox aggregate implementation https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/ApproxDistinctAggregate.cpp#L377-L379, it suggests `addRawInput() method would use DecodedVector’s to decode the input data. Then, loop over rows to update the accumulators. It is a good practice to define a member variable of type DecodedVector for each input vector. This allows for reusing the memory needed to decode the inputs between batches of input.` But @zuochunwei review velox code and think current solution, one `DecodedVector` for all columns is better, I'm not sure which one is better, so I just approve now, and wait for the benchmark to test the performance and memory usage

@yimin-yang please make a test for it, any question, please contact me

File: cpp/velox/shuffle/VeloxSplitter.cc [hunk 3]
```diff
@@ -365,115 +364,30 @@ arrow::Status VeloxSplitter::SplitFixedWidthValueBuffer(const velox::RowVector&
...
       case 16:
-        RETURN_NOT_OK(SplitFixedType<uint16_t>(src_addr, dst_addrs));
+        RETURN_NOT_OK(SplitFixedType<uint16_t>(column, dst_addrs));
         break;
       case 32:
-        RETURN_NOT_OK(SplitFixedType<uint32_t>(src_addr, dst_addrs));
+        RETURN_NOT_OK(SplitFixedType<uint32_t>(column, dst_addrs));
         break;
       case 64:
-#ifdef PROCESSAVX
```
Comment (jinchengchenghh) [inline cpp/velox/shuffle/VeloxSplitter.cc:384]: When we will define `PROCESSAVX`? @FelixYBW How can I test performance difference with `PROCESSAVX`

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxSplitter.cc:384]: I found that the macro `PROCESSAVX ` was not defined in `splitter.cc` also, so I remove it now.
I think that it only affects the performance, not the correctness, We can test it in the future.

**Pull request merged**